src/main/scala/org/ergoplatform/nodeView/history/modifierprocessors/UtxoSnapshotChunkProcessor.scala

```diff
@@ -0,0 +1,92 @@
+package org.ergoplatform.nodeView.history.modifierprocessors
+
+import com.google.common.primitives.Ints
+import io.iohk.iodb.ByteArrayWrapper
+import org.ergoplatform.modifiers.ErgoPersistentModifier
+import org.ergoplatform.modifiers.history.Header
+import org.ergoplatform.modifiers.state.{UtxoSnapshot, UtxoSnapshotChunk, UtxoSnapshotManifest}
+import org.ergoplatform.nodeView.history.storage.HistoryStorage
+import org.ergoplatform.settings.{Algos, Constants}
+import scorex.core.ModifierTypeId
+import scorex.core.consensus.History.ProgressInfo
+import scorex.core.utils.ScorexEncoding
+import scorex.util.{ModifierId, ScorexLogging, bytesToId}
+
+import scala.util.{Failure, Try}
+
+trait UtxoSnapshotChunkProcessor extends ScorexLogging with ScorexEncoding {
+
+  protected val historyStorage: HistoryStorage
+
+  protected val PendingManifestIdKey: ByteArrayWrapper
+
+  private val emptyProgressInfo = ProgressInfo[ErgoPersistentModifier](None, Seq.empty, Seq.empty, Seq.empty)
+
+  protected val LastSnapshotAppliedHeightKey: ByteArrayWrapper =
+    ByteArrayWrapper(Array.fill(Constants.HashLength)(UtxoSnapshot.modifierTypeId))
+
+  protected def toDownload(header: Header): Seq[(ModifierTypeId, ModifierId)]
+
+  protected def lastSnapshotAppliedHeight: Option[Int] = historyStorage.getIndex(LastSnapshotAppliedHeightKey)
+    .map(w => Ints.fromByteArray(w.data))
+
+  protected def pendingManifestOpt: Option[UtxoSnapshotManifest] = {
+    historyStorage.getIndex(PendingManifestIdKey).flatMap {
+      case r if r.data.nonEmpty =>
+        historyStorage.modifierById(bytesToId(r.data))
+          .fold[Option[UtxoSnapshotManifest]](None) {
+            case m: UtxoSnapshotManifest => Some(m)
+            case _ => throw new Error("Wrong history index")
```

Comment:
@catena2w, this branch should never be evaluated in normal situation, if this happens then the history indexes are inconsistent and, as I think, we should throw an Error explicitly, instead of just returning `None` (as it has different semantics, `Exception` has different semantics too, as it is usually expected as result of normal program evaluation, but that particular case is result of wrong program state)